rust/cw-contracts/nft-launchpad/src/contract.rs

```diff
@@ -0,0 +1,348 @@
+                    royalty_percentage: collection.royalty_percentage,
+                },
+                mint_periods: collection.mint_periods,
+            })?,
+            funds: vec![],
+            label: format!(
+                "TR721 codeId:{} collectionId:{} symbol:{}",
+                nft_code_id.unwrap(),
+                collection_id,
+                collection.symbol
+            ),
+        };
+
+        // Update instantiating collection id
+        self.instantiating_collection_id
+            .save(ctx.deps.storage, &collection_id)?;
+
+        let submessage = SubMsg::reply_on_success(instantiate_msg, INSTANTIATE_REPLY_ID);
+
+        Ok(Response::new()
+            .add_submessage(submessage)
+            .add_attribute("action", "collection_deployed")
+            .add_attribute("collection_id", collection_id))
+    }
+
+    #[msg(query)]
+    pub fn get_collection_by_id(
+        &self,
+        ctx: QueryCtx,
+        collection_id: String,
+    ) -> StdResult<Collection> {
+        let collection = self.collections.load(ctx.deps.storage, collection_id)?;
+        Ok(collection)
+    }
+
+    #[msg(query)]
+    pub fn get_config(&self, ctx: QueryCtx) -> StdResult<Config> {
+        let config = self.config.load(ctx.deps.storage)?;
+        Ok(config)
+    }
+
+    #[msg(reply)]
+    pub fn handle_reply(&self, ctx: ReplyCtx, msg: Reply) -> Result<Response, ContractError> {
+        let msg_id = msg.id;
+        let storage = ctx.deps.storage;
+
+        if msg_id == INSTANTIATE_REPLY_ID {
+            let resp = parse_reply_instantiate_data(msg).unwrap();
+            let deployed_addr = resp.contract_address;
+
+            // Get instantiating collection id
+            let collection_id = self.instantiating_collection_id.load(storage)?;
+
+            // Update collection states
+            let mut collection = self.collections.load(storage, collection_id.to_owned())?;
+            collection.deployed_address = Some(deployed_addr.clone());
+            self.collections
+                .save(storage, collection_id.to_owned(), &collection)?;
+
+            return Ok(Response::new()
+                .add_attribute("action", "collection_instantiated")
+                .add_attribute("collection_id", collection_id)
+                .add_attribute("collection_addr", deployed_addr));
+        }
+
+        Err(ContractError::UnknownReply { reply_id: msg_id })
+    }
+}
+
+// Types definitions ------------------------------------------------------
+#[cw_serde]
+pub struct Config {
+    pub name: String,
+    pub nft_code_id: Option<u64>,
+    pub supported_networks: Vec<String>,
+    pub launchpad_admin: Addr,
+    pub owner: Addr,
+}
+
+#[cw_serde]
+pub struct ConfigChanges {
+    pub name: String,
+    pub nft_code_id: Option<u64>,
+    pub supported_networks: Vec<String>,
+    pub launchpad_admin: Option<String>,
+    pub owner: Option<String>,
+}
+
+#[cw_serde]
+#[derive(Default)]
+pub struct Collection {
```

Comment:
```suggestion
pub struct CollectionProject {
```

to disambiguate with the actual Collection

Reply:
Hm, ProjectCollection, no ?
It's a "collection data", not a "project data".
Also, it will be embedded into a "project" in db

Reply:
Why not just `LaunchpadCollection` ?

Reply:
because we're already in the `nft-launchpad` context

Reply:
https://github.com/TERITORI/teritori-dapp/pull/1315/commits/59af3a3726aa26a67c01e1fc0d89422fdd9a8e6b#diff-d4c781af0f12e941e4d2d22efdc47c13be5aa19a1351463b10a50e4f67b57aa7L32